source/common/config/unified_mux/grpc_mux_impl.h

```diff
@@ -0,0 +1,263 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <queue>
+
+#include "envoy/common/random_generator.h"
+#include "envoy/common/time.h"
+#include "envoy/common/token_bucket.h"
+#include "envoy/config/grpc_mux.h"
+#include "envoy/config/subscription.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/grpc/status.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "source/common/common/logger.h"
+#include "source/common/common/utility.h"
+#include "source/common/config/api_version.h"
+#include "source/common/config/grpc_stream.h"
+#include "source/common/config/pausable_ack_queue.h"
+#include "source/common/config/unified_mux/delta_subscription_state.h"
+#include "source/common/config/unified_mux/sotw_subscription_state.h"
+#include "source/common/config/watch_map.h"
+#include "source/common/grpc/common.h"
+
+#include "absl/container/node_hash_map.h"
+
+namespace Envoy {
+namespace Config {
+namespace UnifiedMux {
+
+// Manages subscriptions to one or more type of resource. The logical protocol
+// state of those subscription(s) is handled by SubscriptionState.
+// This class owns the GrpcStream used to talk to the server, maintains queuing
+// logic to properly order the subscription(s)' various messages, and allows
+// starting/stopping/pausing of the subscriptions.
+template <class S, class F, class RQ, class RS>
```

Comment:
@htuch: template base class definition here

Reply:
Can you reduce the number of template parameters by preferring subclassing relationship for things we directly control, like "state" classes?

Reply:
I don't think I can: there's no state interface, for instance, as the base state class is a template with template public methods (getNextRequestAckless and getNextRequestWithAck). B/c state and methods that return state are used in the base mux class, the concrete class must be known at compilation time.

I think this is as small a template-based implementation as it can be, and probably the reason why the original implementation returned pointers to void in `getNextRequestAckless` and `getNextRequestWithAck`...

Reply:
I just realised I might be able to get away with adding an empty "marker" interface for state classes, and then cast that to an appropriate concrete state class in mux implementations. WDYT, @htuch?

Reply:
Not super great, but probably preferable to exploding out state variables. Why is base state a template?

Reply:
It basically boils down to DeltaDiscovery[Request/Response]Base and sotw Discovery[Request/Response] not having a common base class. In the original implementation SubscriptionState methods that returned requests/responses or accepted them as parameters were using pointers to void. These have now been changed to templates (here, for example: https://github.com/envoyproxy/envoy/pull/16486/files#diff-ec4538b5bf47eacd7205e605ae59723bbef2a554ac9eae0697a6ab36d1fd1c8fR60).

Reply:
I see. I think maybe consider this a request to do what's possible without gymnastics to minimize templates and template variable count within a template where possible. Do the least bad option ;-)